File: client/src/app/models/moderator-task.ts
import { AbstractModel } from './abstract-model';
import { User } from './user';
import { deserializeLCObject, LCObject, ObjectType } from './object';

export class ModeratorTask extends AbstractModel {
  title: string;
  description: string;
  completed: boolean;
  timeFinished: Date;
  createdBy: User;
  assignedTo: User;
  finishedBy: User;
  object: LCObject;
  objectType: ObjectType;
  objectId: string;

  public static deserialize(payload: any): ModeratorTask {
    const task = new ModeratorTask();
    AbstractModel.deserializeAbstractAttributes(task, payload);
    task.title = payload.title;
    task.description = payload.description;
    task.completed = payload.completed;
    task.timeFinished = payload.timeFinished
      ? new Date(payload.timeFinished)
      : null;
    task.createdBy = payload.createdBy
      ? User.deserialize(payload.createdBy)
      : null;
    task.assignedTo = payload.assignedTo
      ? User.deserialize(payload.assignedTo)
      : null;
    task.finishedBy = payload.finishedBy
      ? User.deserialize(payload.finishedBy)
      : null;
    task.objectType = payload.objectType;
    task.objectId = payload.objectId;
    task.object = payload.object
      ? deserializeLCObject(payload.objectType, payload.object)
      : null;
    return task;
  }

  public static serializeForCreate(task: ModeratorTask): any {
    return {
      title: task.title,
      description: task.description,
      objectType: task.objectType,
      objectId: task.objectId,
      assignedToId: task.assignedTo?.id ?? null,
    };
  }

  public static serializeForUpdate(task: ModeratorTask): any {
    return {
      title: task.title,
      description: task.description,
      assignedToId: task.assignedTo?.id ?? null,
    };
  }
}

File: server/src/models/moderator_task.py
from sqlalchemy import and_
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.event import listens_for
from sqlalchemy.orm import Session
from sqlalchemy_utils import generic_relationship

from extensions import db
from models.area import Area
from models.base_entity import BaseEntity
from models.crag import Crag
from models.line import Line
from models.sector import Sector


class ModeratorTask(BaseEntity):
    __tablename__ = "moderator_tasks"

    title = db.Column(db.String(120), nullable=False)
    description = db.Column(db.Text, nullable=True)
    completed = db.Column(db.Boolean, nullable=False, default=False, server_default="false")
    time_finished = db.Column(db.DateTime(), nullable=True)
    assigned_to_id = db.Column(UUID(), db.ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    assigned_to = db.relationship("User", foreign_keys=[assigned_to_id])
    finished_by_id = db.Column(UUID(), db.ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    finished_by = db.relationship("User", foreign_keys=[finished_by_id])

    object_type = db.Column(db.Unicode(255), nullable=False)
    object_id = db.Column(UUID(), nullable=False)
    object = generic_relationship(object_type, object_id)


@listens_for(Line, "before_delete")
def cascade_delete_line_moderator_tasks(mapper, connection, target):
    session = Session.object_session(target)
    session.query(ModeratorTask).filter(
        and_(ModeratorTask.object_id == target.id, ModeratorTask.object_type == "Line")
    ).delete()


@listens_for(Area, "before_delete")
def cascade_delete_area_moderator_tasks(mapper, connection, target):
    session = Session.object_session(target)
    session.query(ModeratorTask).filter(
        and_(ModeratorTask.object_id == target.id, ModeratorTask.object_type == "Area")
    ).delete()


@listens_for(Sector, "before_delete")
def cascade_delete_sector_moderator_tasks(mapper, connection, target):
    session = Session.object_session(target)
    session.query(ModeratorTask).filter(
        and_(ModeratorTask.object_id == target.id, ModeratorTask.object_type == "Sector")
    ).delete()


@listens_for(Crag, "before_delete")
def cascade_delete_crag_moderator_tasks(mapper, connection, target):
    session = Session.object_session(target)
    session.query(ModeratorTask).filter(
        and_(ModeratorTask.object_id == target.id, ModeratorTask.object_type == "Crag")
    ).delete()
